py/src/braintrust/logger.py

```diff
@@ -1649,6 +1722,46 @@ def _validate_and_sanitize_experiment_log_full_args(event: Mapping[str, Any], ha
     return event
 
 
+def _deep_copy_event(event: Mapping[str, Any]) -> Dict[str, Any]:
+    """
+    Creates a deep copy of the given event. Replaces references to user objects
+    with placeholder strings to ensure serializability, except for `Attachment`
+    objects, which are preserved and not deep-copied.
+    """
+
+    def _deep_copy_object(v: Any) -> Any:
+        if isinstance(v, Span):
+            return "<span>"
+        elif isinstance(v, Experiment):
+            return "<experiment>"
+        elif isinstance(v, Dataset):
+            return "<dataset>"
+        elif isinstance(v, Logger):
+            return "<logger>"
+        elif isinstance(v, Attachment):
+            return v
+        else:
+            # No need to handle primitives explicitly because deepcopy will do
+            # it for us.
+            return copy.deepcopy(v)
+
+    ret: Dict[str, Any] = {}
+
+    for k, v in event.items():
+        k = str(k)
```

Comment:
nit: should we put stuff like this into this function (or just leave it out altogether)? https://github.com/braintrustdata/braintrust-sdk/blob/546c402c6c12aea4a218b63bc4420d8b4c1e37d8/py/src/braintrust/bt_json.py#L6

Reply:
The purpose of this line is to prevent dict keys from holding references to user objects. `json.dumps` will already coerce all dict key values to str, so we don't need to add it to `bt_json`. However, `bt_json` runs at log upload time so we cannot rely on it to remove the references.

I'll leave a comment for this.